application/application.go

```diff
@@ -0,0 +1,585 @@
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	return nil
+}
+
+func PublishVersion(smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	fmt.Println("Publishing")
+
+	// This needs to do the validation on required fields etc.
+	errModel := models.ValidateVersion(versionUpdate)
+	if errModel != nil {
+		fmt.Println("Validation Failed")
+		fmt.Println(errModel)
+		return errModel
+	} else {
+		fmt.Println("Validation passed, continue")
+		data := versionDetails.baseLogData()
+
+		err := UpdateVersionInfo(smDS, ctx, currentVersion, versionUpdate)
+		if err != nil {
+			log.Error(ctx, "State machine - Publish: UpdateVersionInfo : failed to update the version", err, data)
+			return err
+		}
+
+		log.Info(ctx, "attempting to publish edition", data)
+		err = PublishEdition(smDS, ctx, versionUpdate, versionDetails, data)
+		if err != nil {
+			log.Error(ctx, "State machine - Publish: PublishEdition : failed to publish edition", err, data)
+			return err
+		}
+
+		// This is only applicable to CMD datasets
+		if currentVersion.Type == "v4" {
+			err = PublishCMDInstance(smDS, ctx, versionUpdate, data)
+			if err != nil {
+				log.Error(ctx, "State machine - Publish: PublishInstance : failed to publish instance", err, data)
+				return err
+			}
+		}
+
+		err = PublishDataset(smDS, ctx, currentDataset, currentVersion, versionUpdate, versionDetails, data)
+		if err != nil {
+			log.Error(ctx, "State machine - Publish: PublishDataset : failed to publish dataset", err, data)
+			return err
+		}
+
+		return nil
+
+	}
+}
+
+func CompleteVersion(smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	return nil
+}
+
+func DetachVersion(smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	return nil
+}
+
+func FailVersion(smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	return nil
+}
+
+func UpdateVersionInfo(smDS *StateMachineDatasetAPI, ctx context.Context,
```

Comment:
Consider whether this needs to be exported.

Reply:
I've now amended this